data_structures/treap.cpp

```diff
@@ -0,0 +1,228 @@
+
+/**
+ * @file
+ * @brief A balanced binary search tree (BST) on the basis of binary search tree
+ * and heap: the [Treap](https://en.wikipedia.org/wiki/Treap) algorithm
+ * implementation
+ * @details
+ * Implementation of the treap data structre
+ *
+ * Support operations including insert, erase, and query (the rank of specified
+ * element or the element ranked x) as the same as BST
+ *
+ * But these operations take O(log N) time, since treap keeps property of heap
+ * using rotate operation, and the desired depth of the tree is O(log N).
+ * There's very little chance that it will degenerate into a chain like BST
+ *
+ * @author [Kairao ZHENG](https://github.com/fgmn)
+ */
+
+#include <iostream>  /// For IO operations
+
+const int maxNode = 1e6 + 5;
+/**
+ * @brief Struct representation of the treap
+ */
+struct Treap {
+    /**
+     * @brief Basic data definition
+     * root The root of the treap
+     * treapCnt Total number of current nodes in the treap
+     * key Node identifier
+     * priority Random priority
+     * childs [i][0] represents left child of node i, and [i][1] represents
+     * right size Maintains the subtree size for ranking query cnt The number of
+     * copies per node
+     */
+    int root, treapCnt, key[maxNode], priority[maxNode];
+    int childs[maxNode][2], cnt[maxNode], size[maxNode];
```

Comment:
this whole description could be better expressed as documentations to the comments
i.e:
```cpp
int root; //< root of the treap